### ISSUE TITLE
PVC "requests.storage" is being translated into quota

### ISSUE THREAD

**Issue by Ramblurr** [democratic-csi/democratic-csi#47]
Code in question:

https://github.com/democratic-csi/democratic-csi/blob/9156b65b0749d8e3d64c51cc5330f1ab744e1b73/src/driver/controller-zfs-ssh/index.js#L352-L354

During testing I encountered a situation where my pods were unable to write storage when using a PVC such as the following:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-pvc
spec:
  accessModes:
    - ReadWriteMany
  storageClassName: freenas-nfs
  resources:
    requests:
      storage: 1Mi
```

It turns out the requests.storage value is set as the refquota value in truenas. My pods had written more than 1 Mi

It is my understanding that the `requests.storage` value is for a minimum amount of storage, not a maximum. So it doesn't seem right that this would result in a quota of 1Mi.  Or is my understanding of `requests.storage` wrong?

**Comment by travisghansen** [comment]
Having historical context into k8s and storage helps answer this question. Originally there were no storage classes or auto provisioner. PVs were created manually (typically a bunch up front of various sizes) and then a pvc would be matched to one of the already existing PVs. This explains your expectations about getting an ‘at least’ amount of storage as that’s how it worked (incidentally it also gives insight into the reclaim policy idea as the same pv would end up getting uses by multiple pvcs over time so some sort of rm -rf was required to clean it before subsequent use).

Later provisioners came along and even later csi. When provisioners/storage classes hit that field became more of a ‘you get what you ask for’ than an ‘at least’ situation as effectively each pv was now getting uniquely provisioned for each pvc.

I’m not even entirely sure if k8s supports the limit feature of csi yet. I don’t believe so. If you look at the logic for the zvols it reflects some of this weirdness as zvols (based on the requested amount and the blocksize) may have capacity that is slightly larger than the request. I handle this case if a limit is set and that value is above the limit it throws an error.

In short, it’s expected behavior. If you wish to not have limits and just have your full parent dataset size available to all pvcs/pvs then you can disable the refquota feature. Note that will only work for nfs and not block/iscsi as you must give zvols a size.

**Comment by Ramblurr** [comment]
Thanks for the background info @travisghansen !

>  If you wish to not have limits and just have your full parent dataset size available to all pvcs/pvs then you can disable the refquota feature.

This is a TrueNAS level option right?

Is it possible to have some PVCs have no limits while others have limits?  Would this involve creating multiple democratic csi installs using different dataset parent names?

**Comment by travisghansen** [comment]
It's an option per-deploy of `democratic-csi` yes.

- https://github.com/democratic-csi/democratic-csi/blob/master/examples/freenas-nfs.yaml#L51

You would need to deploy `democratic-csi` twice with 2 different storage classes (at least 1 each). Hypothetically this could be done as a configuration at the storage class level which is a good thought to consider but currently it's per-deployment :( In that case you'd have a single deployment of `democratic-csi` but 2 storage class bound to it. The option would be manifested as a storage class `parameter`.

**Comment by Ramblurr** [comment]
Thanks @travisghansen, that clears things up for me.
